### ISSUE TITLE
INTEGRATION_world_remove test fails on Ubuntu

### ISSUE THREAD

**Issue by scpeters** [osrf/gazebo#3077]
The `INTEGRATION_world_remove` test currently passes on macOS but fails on Ubuntu CI, aborting during the `RemoveBlankWorld` test:

* macOS: https://build.osrfoundation.org/job/gazebo-ci-gazebo11-homebrew-amd64/89/testReport/(root)/WorldRemoveTest_WorldRemoveTest/
* macOS: https://build.osrfoundation.org/job/gazebo-ci-gazebo11-homebrew-amd64/89/testReport/(root)/RemoveJointTypes_WorldRemoveJointsTest/
* Ubuntu: https://build.osrfoundation.org/job/gazebo-ci-gazebo11-bionic-amd64-gpu-nvidia/85/testReport/(root)/INTEGRATION_world_remove/test_ran/

Ubuntu console log:

~~~
[ RUN      ] WorldRemoveTest/WorldRemoveTest.RemoveBlankWorld/0
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 172.23.1.220
[Dbg] [ServerFixture.cc:203] ServerFixture load in 0.8 seconds, timeout after 600 seconds
[Msg] Stats before removing world:
- WorldPtr use count: [14]
- PhysicsEnginePtr use count: [4]
- Topics in this world: [15]
INTEGRATION_world_remove: /usr/include/boost/smart_ptr/shared_ptr.hpp:734: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = gazebo::physics::Base; typename boost::detail::sp_member_access<T>::type = gazebo::physics::Base*]: Assertion `px != 0' failed.
Aborted (core dumped)
~~~

Using my email archive of gazebo11 CI, I've pinpointed the failure to e7fdfcd1f16179c7fc0c20cbcc2a17a493eab050, when https://github.com/osrf/gazebo/pull/2734 was merged. Within that pull request, I was further able to track the change down to aa591b20dc0ea0efec08c836842438a8b4968859. Prior to this change, the `gazebo::common::convertPosesToSdf16` would `return` immediately if the original version of the SDFormat file was less than 1.7, but after this change the `gazebo::common::convertPosesToSdf16` does work for all SDFormat files. I suspect that earlier changes are related as well to this test failure.

To confirm this, I checked out f44cce5c58448cd9918ff902728932c2af184c10  (the revision before #2734 was merged) and applied the following patch to increase the sdf version of the `blank.world` file to `1.7`. I was able to reproduce the test failure with this patch alone.

~~~
diff --git a/worlds/blank.world b/worlds/blank.world
index d533e33ab2..0678bf433b 100644
--- a/worlds/blank.world
+++ b/worlds/blank.world
@@ -1,5 +1,5 @@
 <?xml version="1.0" ?>
-<sdf version="1.5">
+<sdf version="1.7">
   <world name="default">
     <scene>
       <ambient>0.2 0.2 0.2 1</ambient>
~~~

I suspect the underlying issue is from either of the following PRs:

* https://osrf-migration.github.io/gazebo-gh-pages/#!/osrf/gazebo/pull-requests/3133/page/1
* https://osrf-migration.github.io/gazebo-gh-pages/#!/osrf/gazebo/pull-requests/3189/page/1

**Comment by scpeters** [comment]
potential fix in #3078

**Comment by scpeters** [comment]
I forgot to post that I had collected some backtraces from failures of `INTEGRATION_world_remove` that showed the failure occurs in a thread running `World::Step` after another thread has nearly finished calling `World::Fini`. This motivated the fix in #3078.

~~~
Thread 86 (Thread 0x7fff767fb700 (LWP 26024)):
#0  0x00007ffff42d0fb7 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff42d2921 in __GI_abort () at abort.c:79
#2  0x00007ffff42c248a in __assert_fail_base (fmt=0x7ffff4449750 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x7ffff72c2d94 "px != 0", file=file@entry=0x7ffff72c2cf8 "/usr/include/boost/smart_ptr/shared_ptr.hpp", line=line@entry=734, function=function@entry=0x7ffff731d840 <_ZZNK5boost10shared_ptrIN6gazebo7physics4BaseEEptEvE19__PRETTY_FUNCTION__> "typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = gazebo::physics::Base; typename boost::detail::sp_member_access<T>::type = gazebo::physics::Base*]") at assert.c:92
#3  0x00007ffff42c2502 in __GI___assert_fail (assertion=assertion@entry=0x7ffff72c2d94 "px != 0", file=file@entry=0x7ffff72c2cf8 "/usr/include/boost/smart_ptr/shared_ptr.hpp", line=line@entry=734, function=function@entry=0x7ffff731d840 <_ZZNK5boost10shared_ptrIN6gazebo7physics4BaseEEptEvE19__PRETTY_FUNCTION__> "typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = gazebo::physics::Base; typename boost::detail::sp_member_access<T>::type = gazebo::physics::Base*]") at assert.c:101
#4  0x00007ffff7293753 in boost::shared_ptr<gazebo::physics::Base>::operator->() const (this=<optimized out>)
    at /usr/include/boost/smart_ptr/shared_ptr.hpp:734
#5  0x00007ffff72946b9 in boost::shared_ptr<gazebo::physics::Base>::operator->() const (this=<optimized out>)
    at /usr/include/boost/smart_ptr/detail/sp_counted_base_std_atomic.hpp:124
#6  0x00007ffff72946b9 in gazebo::physics::World::ModelUpdateSingleLoop() (this=0x7fff429761f0)
    at gazebo/physics/World.cc:1646
#7  0x00007ffff729dd3f in gazebo::physics::World::Update() (this=this@entry=0x7fff429761f0)
    at gazebo/physics/World.cc:804
#8  0x00007ffff72a974b in gazebo::physics::World::Step() (this=this@entry=0x7fff429761f0)
    at gazebo/physics/World.cc:728
#9  0x00007ffff72a9b9d in gazebo::physics::World::RunLoop() (this=0x7fff429761f0)
    at gazebo/physics/World.cc:551
#10 0x00007ffff4cf66df in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007ffff7bbb6db in start_thread (arg=0x7fff767fb700) at pthread_create.c:463
#12 0x00007ffff43b371f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
~~~

~~~
Thread 1 (Thread 0x7ffff7f3fa00 (LWP 25936)):
#0  0x00007ffff7bbcd2d in __GI___pthread_timedjoin_ex (threadid=140735181469440, thread_return=0x0, abstime=0x0, block=<optimized out>) at pthread_join_common.c:89
#1  0x00007ffff4cf6933 in std::thread::join() () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007ffff7299a89 in gazebo::physics::World::Fini() (this=0x7fff429761f0)
    at gazebo/physics/World.cc:986
#3  0x00007ffff726daf2 in gazebo::physics::remove_worlds() ()
    at gazebo/physics/PhysicsIface.cc:181
#4  0x0000555555576abf in WorldRemoveTest::RemoveBlankWorld(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (this=<optimized out>, _physicsEngine=...)
    at test/integration/world_remove.cc:149
#5  0x0000555555578654 in WorldRemoveTest_RemoveBlankWorld_Test::TestBody() (this=0x7fff4170cdb0)
    at test/integration/world_remove.cc:397
#6  0x00005555555c475a in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (location=0x5555555e46c4 "the test body", method=<optimized out>, object=0x7fff4170cdb0)
    at test/gtest/src/gtest.cc:2421
#7  0x00005555555c475a in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (object=object@entry=0x7fff4170cdb0, method=<optimized out>, location=location@entry=0x5555555e46c4 "the test body") at test/gtest/src/gtest.cc:2457
#8  0x00005555555baf1a in testing::Test::Run() (this=0x7fff4170cdb0)
    at test/gtest/src/gtest.cc:2495
#9  0x00005555555bafff in testing::TestInfo::Run() (this=0x555556030900)
    at test/gtest/src/gtest.cc:2671
#10 0x00005555555bb135 in testing::TestCase::Run() (this=0x555555ef3820)
    at test/gtest/src/gtest.cc:2789
#11 0x00005555555bb5d8 in testing::internal::UnitTestImpl::RunAllTests() (this=<optimized out>)
    at test/gtest/src/gtest.cc:5051
#12 0x00005555555c4c6a in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (location=0x5555555e5920 "auxiliary test code (environments or event listeners)", method=<optimized out>, object=0x555555afb290)
    at test/gtest/src/gtest.cc:2421
#13 0x00005555555c4c6a in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (object=0x555555afb290, method=<optimized out>, location=location@entry=0x5555555e5920 "auxiliary test code (environments or event listeners)")
    at test/gtest/src/gtest.cc:2457
#14 0x00005555555bb788 in testing::UnitTest::Run() (this=0x5555558095a0 <testing::UnitTest::GetInstance()::instance>)
    at test/gtest/src/gtest.cc:4667
#15 0x0000555555572f5f in RUN_ALL_TESTS() () at test/gtest/include/gtest/gtest.h:2329
#16 0x0000555555572f5f in main(int, char**) (argc=<optimized out>, argv=<optimized out>)
    at test/integration/world_remove.cc:614
~~~
